tool/teleport-update/main.go

```diff
@@ -0,0 +1,351 @@
+
+	"github.com/gravitational/teleport"
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+	tracehttp "github.com/gravitational/teleport/api/observability/tracing/http"
+	apiutils "github.com/gravitational/teleport/api/utils"
+	libdefaults "github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/modules"
+	libutils "github.com/gravitational/teleport/lib/utils"
+	logutils "github.com/gravitational/teleport/lib/utils/log"
+)
+
+const appHelp = `Teleport Updater
+
+The Teleport Updater updates the version a Teleport agent on a Linux server
+that is being used as agent to provide connectivity to Teleport resources.
+
+The Teleport Updater supports upgrade schedules and automated rollbacks. 
+
+Find out more at https://goteleport.com/docs/updater`
+
+const (
+	// templateEnvVar allows the template for the Teleport tgz to be specified via env var.
+	templateEnvVar = "TELEPORT_URL_TEMPLATE"
+	// proxyServerEnvVar allows the proxy server address to be specified via env var.
+	proxyServerEnvVar = "TELEPORT_PROXY"
+	// updateGroupEnvVar allows the update group to be specified via env var.
+	updateGroupEnvVar = "TELEPORT_UPDATE_GROUP"
+)
+
+const (
+	// cdnURITemplate is the default template for the Teleport tgz download.
+	cdnURITemplate = "https://cdn.teleport.dev/teleport-v{{.Version}}-{{.OS}}-{{.Arch}}-bin.tar.gz"
+)
+
+const (
+	// versionsDirName specifies the name of the subdirectory inside of the Teleport data dir for storing Teleport versions.
+	versionsDirName = "versions"
+	// updatesFileName specifies the name of the file inside versionsDirName containing configuration for the teleport update
+	updatesFileName = "updates.yaml"
+	// lockFileName specifies the name of the file inside versionsDirName containing the flock lock preventing concurrent updater execution.
+	lockFileName = ".lock"
+)
+
+var plog = logutils.NewPackageLogger(teleport.ComponentKey, teleport.ComponentUpdater)
+
+func main() {
+	if err := Run(os.Args[1:]); err != nil {
+		libutils.FatalError(err)
+	}
+}
+
+type cliConfig struct {
+	// Debug logs enabled
+	Debug bool
+	// DataDir for Teleport (usually /var/lib/teleport)
+	DataDir string
+
+	// ProxyServer address, scheme and port optional
+	ProxyServer string
+	// Group identifier for updates (e.g., staging)
+	Group string
+	// Template for the Teleport tgz download URL
+	Template string
+
+	// LogFormat controls the format of logging. Can be either `json` or `text`.
+	// By default, this is `text`.
+	LogFormat string
+}
+
+func Run(args []string) error {
+	var ccfg cliConfig
+	ctx := context.Background()
+	ctx, _ = signal.NotifyContext(ctx, os.Interrupt, syscall.SIGTERM)
+
+	app := libutils.InitCLIParser("teleport-updater", appHelp).Interspersed(false)
+	app.Flag("debug", "Verbose logging to stdout.").Short('d').BoolVar(&ccfg.Debug)
+	app.Flag("data-dir", "Directory to store teleport versions. Access to this directory should be limited.").StringVar(&ccfg.DataDir)
```

Comment:
From an end-user perspective, this should always be the equivalent of `/var/lib/teleport`. Users shouldn't have to worry about multiple data directories. The directory is shared with Teleport, because there is (limited) interaction between the tools (e.g., DB backups, host UUID)

Additionally, the data directory contains more than just the config file (e.g., teleport versions, DB backups, DB metadata files).